services/dkg/pedersen/handler.go

```diff
@@ -127,36 +142,98 @@ mainSwitch:
 				"call setup() first?")
 		}
 
-		isShuffled, err := h.checkIsShuffled(msg.K, msg.C, msg.GetElectionId())
+		shuffleInstances, err := h.getShuffleIfValid(msg.GetElectionId())
 		if err != nil {
-			return xerrors.Errorf("failed to check if the ciphertext has been shuffled: %v", err)
+			return xerrors.Errorf("failed to check if the shuffle is over: %v", err)
 		}
 
-		if !isShuffled {
-			return xerrors.Errorf("the ciphertext has not been shuffled")
-		}
+		publicShares := make([][]etypes.PubShare, len(shuffleInstances))
+		numberOfShuffles := len(shuffleInstances)
 
 		// TODO: check if started before
 		h.RLock()
-		S := suite.Point().Mul(h.privShare.V, msg.K)
-		h.RUnlock()
 
-		partial := suite.Point().Sub(msg.C, S)
+		for _, ballot := range shuffleInstances[numberOfShuffles-1].ShuffledBallots {
+			ballotShares := make([]etypes.PubShare, len(ballot))
+
+			for _, ciphertext := range ballot {
+				K, C, err := ciphertext.GetPoints()
+				if err != nil {
+					return xerrors.Errorf(
+						"could not retrieve a ballot's ElGamal pairs: %v", err)
+				}
+				S := suite.Point().Mul(h.privShare.V, K)
+				partialVal := suite.Point().Sub(C, S)
+
+				partialValMarshalled, err := partialVal.MarshalBinary()
+				if err != nil {
+					return xerrors.Errorf("could not marshal pubShare: %v", err)
+				}
+
+				ballotShares = append(ballotShares, etypes.PubShare{
+					I: h.privShare.I,
+					V: partialValMarshalled,
+				})
+			}
+
+			publicShares = append(publicShares, ballotShares)
+		}
 
-		h.RLock()
-		decryptReply := types.NewDecryptReply(
-			// TODO: check if using the private index is the same as the public
-			// index.
-			int64(h.privShare.I),
-			partial,
-		)
 		h.RUnlock()
 
-		errs := out.Send(decryptReply, from)
-		err = <-errs
-		if err != nil {
-			return xerrors.Errorf("got an error while sending the decrypt "+
-				"reply: %v", err)
+		// loop until our transaction has been accepted, or enough nodes
+		// submitted their pubShares
+		for {
+			election, err := getElection(h.context, msg.GetElectionId(), h.service)
+			if err != nil {
+				return xerrors.Errorf("could not get the election: %v", err)
+			}
+
+			//TODO: Works with current "shuffleThreshold", but the shuffle threshold
+			// should be smaller in theory ? (1/3 + 1 vs 2/3 + 1 ? )
+			differentPubShares := len(election.PubSharesArchive.PubSharesSubmissions)
+			if differentPubShares >= election.ShuffleThreshold {
```

Comment:
yeah, now we should have a generic `securityThreshold` as the election attribute, which denotes the shuffle and decryption threshold.

Reply:
Yes, but why do we currently use 2/3 + 1 as a shuffle treshold? If 1/3 + 1 make a shuffle then at least one of the shuffles is correct, right?

Reply:
You're right. I need to meditate on that 😄